controlplane/kubeadm/internal/control_plane.go

```diff
@@ -31,17 +32,25 @@ import (
 // ControlPlane holds business logic around control planes.
 // It should never need to connect to a service, that responsibility lies outside of this struct.
 type ControlPlane struct {
-	KCP      *controlplanev1.KubeadmControlPlane
-	Cluster  *clusterv1.Cluster
-	Machines FilterableMachineCollection
+	KCP            *controlplanev1.KubeadmControlPlane
+	Cluster        *clusterv1.Cluster
+	Machines       FilterableMachineCollection
+	MachineConfigs map[string]*bootstrapv1.KubeadmConfig
+	InfraObjects   map[string]*unstructured.Unstructured
+
+	// ReconciliationTime is the time of the current reconciliation, and should be used for all "now" calculations
+	ReconciliationTime metav1.Time
 }
 
 // NewControlPlane returns an instantiated ControlPlane.
-func NewControlPlane(cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, ownedMachines FilterableMachineCollection) *ControlPlane {
+func NewControlPlane(cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, ownedMachines FilterableMachineCollection, machineConfigs map[string]*bootstrapv1.KubeadmConfig, infraObjects map[string]*unstructured.Unstructured) *ControlPlane {
```

Comment:
This signature is getting a little unwieldy but I think making an input struct is the wrong decision. I was thinking about making an initializer that takes a `client` and does the fetching, but decided against that for now.